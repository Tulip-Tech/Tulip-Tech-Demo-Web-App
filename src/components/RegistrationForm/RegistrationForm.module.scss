@use 'src/styles/variables';
@use 'src/styles/theme';

.title {
  margin-bottom: 24px;
  font-weight: var(--body-font-weight-bold);
  font-size: 24px;
}

.continue {
  margin: variables.$base-spacing 0;
}

.bottom {
  padding: variables.$base-spacing 0;
  text-align: center;
}

.alreadyAccount {
  font-size: 16px;
}

.login {
  margin-left: 6px;
  font-weight: var(--body-font-weight-bold);
  cursor: pointer;

  &:hover,
  &:focus {
    text-decoration: underline;
  }
}
