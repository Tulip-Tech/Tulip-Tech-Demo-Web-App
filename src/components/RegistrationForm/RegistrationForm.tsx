import React from 'react';
import { useLocation } from 'react-router';
import { useTranslation } from 'react-i18next';
import DOMPurify from 'dompurify';

import styles from './RegistrationForm.module.scss';

import TextField from '#components/TextField/TextField';
import Button from '#components/Button/Button';
import IconButton from '#components/IconButton/IconButton';
import Visibility from '#src/icons/Visibility';
import VisibilityOff from '#src/icons/VisibilityOff';
import PasswordStrength from '#components/PasswordStrength/PasswordStrength';
import Checkbox from '#components/Checkbox/Checkbox';
import FormFeedback from '#components/FormFeedback/FormFeedback';
import LoadingOverlay from '#components/LoadingOverlay/LoadingOverlay';
import Link from '#components/Link/Link';
import { testId } from '#src/utils/common';
import useToggle from '#src/hooks/useToggle';
import { addQueryParam } from '#src/utils/location';
import type { FormErrors } from '#types/form';
import type { RegistrationFormData, Consent } from '#types/account';

type Props = {
  onSubmit: React.FormEventHandler<HTMLFormElement>;
  onChange: React.ChangeEventHandler<HTMLInputElement | HTMLTextAreaElement>;
  onBlur: React.FocusEventHandler<HTMLInputElement | HTMLTextAreaElement>;
  onConsentChange: React.ChangeEventHandler<HTMLInputElement>;
  errors: FormErrors<RegistrationFormData>;
  values: RegistrationFormData;
  loading: boolean;
  consentValues: Record<string, boolean>;
  consentErrors: string[];
  submitting: boolean;
  canSubmit: boolean;
  publisherConsents?: Consent[];
};

const RegistrationForm: React.FC<Props> = ({
  onSubmit,
  onChange,
  onBlur,
  values,
  errors,
  submitting,
  loading,
  canSubmit,
  publisherConsents,
  consentValues,
  onConsentChange,
  consentErrors,
}: Props) => {
  const [viewPassword, toggleViewPassword] = useToggle();

  const { t } = useTranslation('account');
  const location = useLocation();

  const formatConsentLabel = (label: string): string | JSX.Element => {
    const sanitizedLabel = DOMPurify.sanitize(label);
    const hasHrefOpenTag = /<a(.|\n)*?>/.test(sanitizedLabel);
    const hasHrefCloseTag = /<\/a(.|\n)*?>/.test(sanitizedLabel);

    if (hasHrefOpenTag && hasHrefCloseTag) {
      return <span dangerouslySetInnerHTML={{ __html: label }} />;
    }

    return label;
  };

  if (loading) {
    return (
      <div style={{ height: 400 }}>
        <LoadingOverlay inline />
      </div>
    );
  }

  return (
    <form onSubmit={onSubmit} data-testid={testId('registration-form')} noValidate>
      <h2 className={styles.title}>{t('registration.sign_up')}</h2>
      {errors.form ? <FormFeedback variant="error">{errors.form}</FormFeedback> : null}
      <TextField
        value={values.email}
        onChange={onChange}
        onBlur={onBlur}
        label={t('registration.email')}
        placeholder={t('registration.email')}
        error={!!errors.email || !!errors.form}
        helperText={errors.email}
        name="email"
        type="email"
        required
      />
      <TextField
        value={values.password}
        onChange={onChange}
        onBlur={onBlur}
        label={t('registration.password')}
        placeholder={t('registration.password')}
        error={!!errors.password || !!errors.form}
        helperText={
          <React.Fragment>
            <PasswordStrength password={values.password} />
            {t('registration.password_helper_text')}
          </React.Fragment>
        }
        name="password"
        type={viewPassword ? 'text' : 'password'}
        rightControl={
          <IconButton aria-label={viewPassword ? t('registration.hide_password') : t('registration.view_password')} onClick={() => toggleViewPassword()}>
            {viewPassword ? <Visibility /> : <VisibilityOff />}
          </IconButton>
        }
        required
      />
      {publisherConsents?.map((consent, index) => (
        <Checkbox
          key={index}
          name={consent.name}
          value={consent.value || ''}
          error={consentErrors?.includes(consent.name)}
          helperText={consentErrors?.includes(consent.name) ? t('registration.consent_required') : undefined}
          required={consent.required}
          checked={consentValues[consent.name] || false}
          onChange={onConsentChange}
          label={formatConsentLabel(consent.label)}
        />
      ))}
      <Button
        className={styles.continue}
        type="submit"
        label={t('registration.continue')}
        variant="contained"
        color="primary"
        size="large"
        disabled={submitting || !canSubmit}
        fullWidth
      />
      <p className={styles.bottom}>
        {t('registration.already_account')} <Link to={addQueryParam(location, 'u', 'login')}>{t('login.sign_in')}</Link>
      </p>
      {submitting && <LoadingOverlay transparentBackground inline />}
    </form>
  );
};

export default RegistrationForm;
