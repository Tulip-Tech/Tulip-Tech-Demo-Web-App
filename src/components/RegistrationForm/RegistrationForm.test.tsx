import React from 'react';

import RegistrationForm from './RegistrationForm';

import { renderWithRouter } from '#test/testUtils';

describe('<RegistrationForm>', () => {
  test('renders and matches snapshot', () => {
    const { container } = renderWithRouter(
      <RegistrationForm
        onSubmit={vi.fn()}
        onChange={vi.fn()}
        onBlur={vi.fn()}
        values={{ email: '', password: '' }}
        errors={{}}
        submitting={false}
        consentErrors={[]}
        consentValues={{}}
        loading={false}
        canSubmit={true}
        onConsentChange={vi.fn()}
      />,
    );

    expect(container).toMatchSnapshot();
  });
});
