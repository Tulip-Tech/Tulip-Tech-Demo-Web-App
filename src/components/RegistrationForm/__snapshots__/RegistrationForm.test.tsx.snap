// Vitest Snapshot v1

exports[`<RegistrationForm> > renders and matches snapshot 1`] = `
<div>
  <form
    data-testid="registration-form"
    novalidate=""
  >
    <h2
      class="_title_a472a0"
    >
      registration.sign_up
    </h2>
    <div
      class="_textField_e16c1b"
    >
      <label
        class="_label_e16c1b"
        for="text-field_1235_email"
      >
        registration.email
      </label>
      <div
        class="_container_e16c1b"
      >
        <input
          class="_input_e16c1b"
          id="text-field_1235_email"
          name="email"
          placeholder="registration.email"
          required=""
          type="email"
          value=""
        />
      </div>
    </div>
    <div
      class="_textField_e16c1b _rightControl_e16c1b"
    >
      <label
        class="_label_e16c1b"
        for="text-field_1235_password"
      >
        registration.password
      </label>
      <div
        class="_container_e16c1b"
      >
        <input
          class="_input_e16c1b"
          id="text-field_1235_password"
          name="password"
          placeholder="registration.password"
          required=""
          type="password"
          value=""
        />
        <div
          class="_control_e16c1b"
        >
          <div
            aria-label="registration.view_password"
            class="_iconButton_0fef65"
            role="button"
            tabindex="0"
          >
            <svg
              aria-hidden="true"
              class="_icon_1e9999"
              focusable="false"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g>
                <path
                  d="M11.83,9L15,12.16C15,12.11 15,12.05 15,12A3,3 0 0,0 12,9C11.94,9 11.89,9 11.83,9M7.53,9.8L9.08,11.35C9.03,11.56 9,11.77 9,12A3,3 0 0,0 12,15C12.22,15 12.44,14.97 12.65,14.92L14.2,16.47C13.53,16.8 12.79,17 12,17A5,5 0 0,1 7,12C7,11.21 7.2,10.47 7.53,9.8M2,4.27L4.28,6.55L4.73,7C3.08,8.3 1.78,10 1,12C2.73,16.39 7,19.5 12,19.5C13.55,19.5 15.03,19.2 16.38,18.66L16.81,19.08L19.73,22L21,20.73L3.27,3M12,7A5,5 0 0,1 17,12C17,12.64 16.87,13.26 16.64,13.82L19.57,16.75C21.07,15.5 22.27,13.86 23,12C21.27,7.61 17,4.5 12,4.5C10.6,4.5 9.26,4.75 8,5.2L10.17,7.35C10.74,7.13 11.35,7 12,7Z"
                />
              </g>
            </svg>
          </div>
        </div>
      </div>
      <div
        class="_helperText_da58e0"
      >
        registration.password_helper_text
      </div>
    </div>
    <button
      aria-disabled="false"
      class="_button_f8f296 _continue_a472a0 _primary_f8f296 _fullWidth_f8f296 _large_f8f296"
      type="submit"
    >
      <span>
        registration.continue
      </span>
    </button>
    <p
      class="_bottom_a472a0"
    >
      registration.already_account
       
      <a
        class="_link_a14c64"
        href="/?u=login"
      >
        login.sign_in
      </a>
    </p>
  </form>
</div>
`;
